service/windows/src/main.rs

```diff
@@ -0,0 +1,312 @@
+#[macro_use]
+extern crate windows_service;
+
+mod config;
+mod manager;
+mod node;
+mod subprocess;
+
+use crate::config::{get_config_file_path, load_config};
+use crate::subprocess::create_console;
+use anyhow::{bail, Context};
+use log::*;
+use retain_mut::RetainMut;
+use std::ffi::OsString;
+use std::string::String;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+use std::time::{Duration, Instant};
+use std::{env, thread};
+use winapi::shared::winerror::ERROR_FAILED_SERVICE_CONTROLLER_CONNECT;
+use windows_service::{service::*, service_control_handler::*, service_dispatcher};
+
+const SERVICE_NAME: &str = "ConcordiumNode";
+const EVENT_LOG_NAME: &str = "Concordium Node Runner Service";
+
+// Produce an FFI wrapper for the service main function.
+define_windows_service!(ffi_service_main, runner_service_main);
+
+/// Service main entrypoint. This is invoked by the service control dispatcher when the service
+/// is started.
+///
+/// This delegates to run_service, and panics, logging an error, if an unrecoverable error occurs.
+fn runner_service_main(arguments: Vec<OsString>) {
+    winlog::init(EVENT_LOG_NAME).unwrap();
+
+    if let Err(e) = run_service(arguments) {
+        error!(
+            "The node runner service failed for the following reason:\n{:#}",
+            e
+        );
+        panic!()
+    }
+}
+
+/// Macro for constructing a simple status message given the new state and enabled controls (if
+/// any).
+macro_rules! simple_status {
+    ($state: expr, $controls: expr) => {
+        ServiceStatus {
+            service_type: ServiceType::OWN_PROCESS,
+            current_state: $state,
+            controls_accepted: $controls,
+            exit_code: ServiceExitCode::Win32(0),
+            checkpoint: 0,
+            wait_hint: Duration::from_secs(0),
+            process_id: None,
+        }
+    };
+    ($state: expr) => {
+        simple_status!($state, ServiceControlAccept::empty())
+    };
+}
+
+/// Macro for constructing a pending status message given the new state and checkpoint and wait
+/// hint values.
+macro_rules! pending_status {
+    ($state: expr, $checkpoint: expr, $wait_hint: expr) => {
+        ServiceStatus {
+            service_type: ServiceType::OWN_PROCESS,
+            current_state: $state,
+            controls_accepted: ServiceControlAccept::empty(),
+            exit_code: ServiceExitCode::Win32(0),
+            checkpoint: $checkpoint,
+            wait_hint: $wait_hint,
+            process_id: None,
+        }
+    };
+}
```

Comment:
And this can be replaced by a function directly.

Reply:
Done.